source/extensions/filters/http/grpc_web/grpc_web_filter.cc

```diff
@@ -54,6 +74,71 @@ bool GrpcWebFilter::isGrpcWebRequest(const Http::RequestHeaderMap& headers) {
   return false;
 }
 
+bool GrpcWebFilter::hasProtoEncodedGrpcWebContentType(
+    const Http::RequestOrResponseHeaderMap& headers) const {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type != nullptr) {
+    absl::string_view content_type_value = content_type->value().getStringView();
+    // We ignore "parameter" value. Note that "*( ";" parameter )" indicates that there can be
+    // multiple parameters.
+    absl::string_view current_content_type =
+        StringUtil::rtrim(content_type_value.substr(0, content_type_value.find_first_of(';')));
+    // We expect only proto encoding response
+    // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md. And the value of media-type is
+    // case-sensitive https://tools.ietf.org/html/rfc2616#section-3.7.
+    return StringUtil::CaseInsensitiveCompare()(
+               current_content_type, Http::Headers::get().ContentTypeValues.GrpcWebProto) ||
+           StringUtil::CaseInsensitiveCompare()(current_content_type,
+                                                Http::Headers::get().ContentTypeValues.GrpcWeb);
+  }
+  return false;
+}
+
+// If response headers do not contain gRPC or gRPC-Web response headers, it needs transformation.
+bool GrpcWebFilter::needsTransformationForNonProtoEncodedResponse(Http::ResponseHeaderMap& headers,
+                                                                  bool end_stream) const {
+  return Runtime::runtimeFeatureEnabled(
+             "envoy.reloadable_features.grpc_web_fix_non_proto_encoded_response_handling") &&
+         !Grpc::Common::isGrpcResponseHeaders(headers, end_stream) &&
+         !(Http::Utility::getResponseStatus(headers) == enumToInt(Http::Code::OK) &&
```

Comment:
mind commenting why we do not convert if the response code is OK?

Reply:
Sure. Commented.